packages/cli/src/gen-interfaces.ts

```diff
@@ -0,0 +1,165 @@
+    })
+    const rets = funcSig.returnTypes.map((type) => getType(type, structNames, contractNames))
+    const result = `
+      ${annotation}
+      pub fn ${funcSig.name}(${params.join(', ')}) -> (${rets.join(', ')})
+    `
+    functions.push(result.trim())
+    publicFuncIndex += 1
+  })
+  const interfaceDef = format(
+    `@using(methodSelector = false)
+    Interface ${interfaceName} {
+      ${functions.join('\n\n')}
+    }`,
+    3
+  )
+  return { name: interfaceName, def: interfaceDef }
+}
+
+function getType(typeName: string, structNames: string[], contractNames: string[]): string {
+  if (PrimitiveTypes.includes(typeName)) return typeName
+  if (typeName.startsWith('[')) {
+    const [baseType, size] = decodeArrayType(typeName)
+    return `[${getType(baseType, structNames, contractNames)}; ${size}]`
+  }
+  if (structNames.includes(typeName)) return typeName
+  if (contractNames.includes(typeName)) return `I${typeName}`
+  // We currently do not generate artifacts for interface types, so when a function
+  // param/ret is of an interface type, we use `ByteVec` as the param/ret type
+  return 'ByteVec'
+}
+
+function genStructs(structs: Struct[], structNames: string[], contractNames: string[]) {
+  const structDefs = structs.map((s) => {
+    const fields = s.fieldNames.map((fieldName, index) => {
+      const fieldType = getType(s.fieldTypes[`${index}`], structNames, contractNames)
+      const isMutable = s.isMutable[`${index}`]
+      return isMutable ? `mut ${fieldName}: ${fieldType}` : `${fieldName}: ${fieldType}`
+    })
+    return format(
+      `struct ${s.name} {
+        ${fields.join(',\n')}
+      }`,
+      2
+    )
+  })
+  return structDefs.join('\n\n')
+}
+
+function format(str: string, lineToIndentFrom: number): string {
+  const padding = '  ' // 2 spaces
+  const lines = str.trim().split('\n')
+  return lines
+    .map((line, index) => {
+      const newLine = line.trim()
+      if (index < lineToIndentFrom - 1 || index === lines.length - 1) {
+        return newLine
+      } else if (newLine.length === 0) {
+        return line
+      } else {
+        return padding + newLine
+      }
+    })
+    .join('\n')
+}
+
+async function loadContracts(artifactDir: string, structs: Struct[]) {
+  const contracts: Contract[] = []
+  const load = async function (dirPath: string): Promise<void> {
+    const dirents = await fsPromises.readdir(dirPath, { withFileTypes: true })
+    for (const dirent of dirents) {
+      if (dirent.isFile()) {
+        const artifactPath = path.join(dirPath, dirent.name)
+        const contract = await getContractFromArtifact(artifactPath, structs)
+        if (contract !== undefined) contracts.push(contract)
+      } else {
+        const newPath = path.join(dirPath, dirent.name)
+        await load(newPath)
+      }
+    }
+  }
+  await load(artifactDir)
+  return contracts
+}
+
+async function getContractFromArtifact(filePath: string, structs: Struct[]): Promise<Contract | undefined> {
```

Comment:
1. Since all files are generated and should not be modified, maybe we prefix the file name with`__`. E.g. `__IFoo.ral`
2. Let's also add a comment `/// Autogenerated file. Do not edit manually.`